src/core/ws/WebSocketClientManager.ts

```diff
@@ -0,0 +1,211 @@
+   * using the shared `BroadcastChannel` instance.
+   */
+  private notifyOthersAboutDatabaseUpdate(): void {
+    this.channel.postMessage({ type: 'db:update' })
+  }
+
+  private async addClient(client: WebSocketClientConnection): Promise<void> {
+    await this.store.add(client)
+    // Sync the in-memory clients in this runtime with the
+    // updated database. This pulls in all the stored clients.
+    await this.flushDatabaseToMemory()
+    this.notifyOthersAboutDatabaseUpdate()
+  }
+
+  /**
+   * Adds the given `WebSocket` client connection to the set
+   * of all connections. The given connection is always the complete
+   * connection object because `addConnection()` is called only
+   * for the opened connections in the same runtime.
+   */
+  public async addConnection(client: WebSocketClientConnection): Promise<void> {
+    // Store this client in the map of clients created in this runtime.
+    // This way, the manager can distinguish between this runtime clients
+    // and extraneous runtime clients when synchronizing clients storage.
+    this.runtimeClients.set(client.id, client)
+
+    // Add the new client to the storage.
+    await this.addClient(client)
+
+    // Handle the incoming BroadcastChannel messages from other runtimes
+    // that attempt to control this runtime (via a remote connection wrapper).
+    // E.g. another runtime calling `client.send()` for the client in this runtime.
+    const handleExtraneousMessage = (
+      message: MessageEvent<WebSocketBroadcastChannelMessage>,
+    ) => {
+      const { type, payload } = message.data
+
+      // Ignore broadcasted messages for other clients.
+      if (
+        typeof payload === 'object' &&
+        'clientId' in payload &&
+        payload.clientId !== client.id
+      ) {
+        return
+      }
+
+      switch (type) {
+        case 'extraneous:send': {
+          client.send(payload.data)
+          break
+        }
+
+        case 'extraneous:close': {
+          client.close(payload.code, payload.reason)
+          break
+        }
+      }
+    }
+
+    const abortController = new AbortController()
+
+    this.channel.addEventListener('message', handleExtraneousMessage, {
+      signal: abortController.signal,
+    })
+
+    // Once closed, this connection cannot be operated on.
+    // This must include the extraneous runtimes as well.
+    client.addEventListener('close', () => abortController.abort(), {
+      once: true,
+    })
+  }
+}
+
+/**
+ * A wrapper class to operate with WebSocket client connections
+ * from other runtimes. This class maintains 1-1 public API
+ * compatibility to the `WebSocketClientConnection` but relies
+ * on the given `BroadcastChannel` to communicate instructions
+ * with the client connections from other runtimes.
+ */
+export class WebSocketRemoteClientConnection
+  implements WebSocketClientConnectionProtocol
+{
+  constructor(
+    public readonly id: string,
+    public readonly url: URL,
```

Comment:
question: Why is the `url` necessary? This is not used.

Reply:
I've got this removed already. Likely forgot to push :)